nexus/src/db/datastore.rs

```diff
@@ -3394,6 +3396,93 @@ impl DataStore {
                 ))
             })
     }
+
+    pub async fn allocate_static_v6_address(
+        &self,
+        allocation_config: &StaticV6AllocationConfig,
+    ) -> Result<Ipv6Addr, Error> {
+        use db::schema::static_v6_address::dsl;
+
+        let allocation_config = allocation_config.clone();
+
+        self.pool()
+            .transaction(move |conn| {
+                // First, search for released addresses
+                let result: Result<StaticV6Address, diesel::result::Error> =
+                    dsl::static_v6_address
+                        .filter(dsl::allocated.eq(false))
+                        .limit(1)
+                        .get_result(conn);
+
+                if let Ok(row) = result {
+                    diesel::update(dsl::static_v6_address)
+                        .filter(dsl::address.eq(&row.address))
+                        .set(dsl::allocated.eq(true))
+                        .execute(conn)?;
+
+                    return Ok(row.address.parse().unwrap());
+                }
+
+                // If no released addresses exist, allocate a new one
+                // XXX there's got to be a better way! store the iterator
+                // somewhere?
+                for ip in allocation_config.subnet.iter() {
+                    // Do not allocate less than the minimum address
+                    if ip <= allocation_config.minimum_address {
+                        continue;
+                    }
+
+                    let allocated_ip = StaticV6Address {
+                        address: ip.to_string(),
+                        allocated: true,
+                    };
+
+                    let result = diesel::insert_into(dsl::static_v6_address)
+                        .values(allocated_ip)
+                        .returning(StaticV6Address::as_returning())
+                        .get_result(conn);
+
+                    match result {
+                        Ok(allocated_address) => {
+                            return Ok(allocated_address
+                                .address
+                                .parse()
+                                .unwrap());
+                        }
+
+                        Err(_) => {
+                            // XXX should match on primary key conflict
```

Comment:
@bnaecker 's subnet allocation has an example of this:

https://github.com/oxidecomputer/omicron/blob/b259f45509bb9a5ba0bb2b656720d1d0472bc6e0/nexus/src/db/subnet_allocation.rs#L568-L573

Reply:
@jmpesp If you find yourself checking specific database error conditions or messages, let me know. I've been prototyping some tooling to make that kind of thing easier. That hasn't been too valuable yet, with only 2 use sites, but if we add more...